### PR TITLE
allow_list for allowing only specified addrs to get randomness

### PR DESCRIPTION
**Pull request by dgabriele**
- Used Map keys to represent allow_list.
- Query method for checking if a contract addr is "allowed"
- Execute method for updating allow_list
- Optional setting of allow_list in instantiation
- Added allowlist "enabled" flag to config
- New unit tests

### CODE REVIEW

File: contracts/nois-proxy/src/contract.rs [hunk 1]
```diff
@@ -723,7 +767,9 @@ mod tests {
 
     const CREATOR: &str = "creator";
 
-    fn setup() -> OwnedDeps<MockStorage, MockApi, MockQuerier> {
+    fn setup(
+        instantiate_msg: Option<InstantiateMsg>,
```
Comment (dgabriele) [inline contracts/nois-proxy/src/contract.rs:771]: Added the ability to pass in a custom `InstantiateMsg`, but if None provided, defaults to the existing hard-coded one.

**Comment by kaisbaccour** [comment]
lgtm, let's see what @webmaster128 thinks

**Comment by webmaster128** [comment]
Beautiful PR, thank you.

But let's store the allow list as a separate `ALLOW_LIST: Map<&Addr, u8>` where you save any value that is short in JSON (e.g. `1u8`) as the value. An `ALLOW_LIST.has(..., &info.sender)` checks the existence in a single db read.

Loading and deserializing a potentially long JSON list of addresses can easily become costy. Especially since it is loaded every time we need a `Config`.

Keeping `allow_list_enabled: bool` in the config as an alternative to the current `Option` makes sense. `InstantiteMsg` looks good as is.

**Comment by dgabriele** [comment]
> Beautiful PR, thank you.
> 
> But let's store the allow list as a separate `ALLOW_LIST: Map<&Addr, u8>` where you save any value that is short in JSON (e.g. `1u8`) as the value. An `ALLOW_LIST.has(..., &info.sender)` checks the existence in a single db read.
> 
> Loading and deserializing a potentially long JSON list of addresses can easily become costy. Especially since it is loaded every time we need a `Config`.
> 
> Keeping `allow_list_enabled: bool` in the config as an alternative to the current `Option` makes sense. `InstantiteMsg` looks good as is.

Indeed. I started to do that originally but changed my mind, thinking you might like it better if I kept it all in the config :). I'll put it in a map

File: contracts/nois-proxy/src/msg.rs [hunk 2]
```diff
@@ -146,6 +161,11 @@ pub struct GatewayChannelResponse {
     pub channel: Option<String>,
 }
 
+#[cw_serde]
+pub struct IsAllowedResponse {
+    pub is_allowed: bool,
```
Comment (webmaster128) [inline contracts/nois-proxy/src/msg.rs:166]:
```suggestion
    /// Returns true if and only if the address is allowed to use the proxy.
    /// This is true if no allow-list is in place or the given address is allow-listed.
    pub is_allowed: bool,
```

File: contracts/nois-proxy/src/contract.rs [hunk 3]
```diff
@@ -104,6 +118,9 @@ pub fn execute(
             amount,
             address,
         } => execute_withdraw(deps, env, info, denom, amount, address),
+        ExecuteMsg::UpdateAllowList { add, remove } => {
+            execute_update_allowlist(deps, env, info, &add, &remove)
```
Comment (webmaster128) [inline contracts/nois-proxy/src/contract.rs:122]: Let's pass those vectors owned. Maybe this is useful later on.

```suggestion
            execute_update_allowlist(deps, env, info, add, remove)
```

File: contracts/nois-proxy/src/contract.rs [hunk 4]
```diff
@@ -257,6 +289,35 @@ fn execute_withdraw(
     withdraw_unchecked(deps, env, denom, amount, address)
 }
 
+fn execute_update_allowlist(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    added_addresses: &Vec<String>,
+    removed_addresses: &Vec<String>,
```
Comment (webmaster128) [inline contracts/nois-proxy/src/contract.rs:297]:
```suggestion
    added_addresses: Vec<String>,
    removed_addresses: Vec<String>,
```

File: contracts/nois-proxy/src/contract.rs [hunk 5]
```diff
@@ -257,6 +289,35 @@ fn execute_withdraw(
...
+) -> Result<Response, ContractError> {
+    update_allowlist(deps, added_addresses, removed_addresses)?;
+    Ok(Response::new().add_attribute("action", "update_allowlist"))
+}
+
+/// Adds and remove entries from the allow list.
+fn update_allowlist(
+    deps: DepsMut,
+    added_addresses: &Vec<String>,
+    removed_addresses: &Vec<String>,
```
Comment (webmaster128) [inline contracts/nois-proxy/src/contract.rs:307]:
```suggestion
    added_addresses: Vec<String>,
    removed_addresses: Vec<String>,
```

File: contracts/nois-proxy/src/msg.rs [hunk 6]
```diff
@@ -43,6 +47,8 @@ pub enum ExecuteMsg {
         /// The amount of tokens the proxy sends for each randomness request to the Nois chain
         nois_beacon_price: Option<Uint128>,
         mode: Option<OperationalMode>,
+        /// Toggle address whitelist allowed to get randomness
+        allowlist_enabled: Option<bool>,
```
Comment (webmaster128) [inline contracts/nois-proxy/src/msg.rs:51]: SudoMsg::SetConfig should get this too

File: contracts/nois-proxy/src/state.rs [hunk 7]
```diff
@@ -47,10 +47,17 @@ pub struct Config {
     /// The time (on the Nois chain) the price info was created
     pub nois_beacon_price_updated: Timestamp,
     pub mode: OperationalMode,
+    /// Enable whitelist of addresses allowed to get randomness
+    pub allowlist_enabled: bool,
```
Comment (webmaster128) [inline contracts/nois-proxy/src/state.rs:51]:
```suggestion
    /// Enable whitelist of addresses allowed to get randomness.
    /// This is an Option for compatibility with older versions of the contract. If set to None it means disabled.
    pub allowlist_enabled: Option<bool>,
```

**Review by webmaster128 — COMMENTED** [review]
Good stuff. I think we can make this a compatible addition when the config field becomes an Option.

File: contracts/nois-proxy/src/msg.rs (hunk 2 above)
Comment (webmaster128) [reply contracts/nois-proxy/src/msg.rs:166]: Actually I am not sure how the query should work. I think what you implemented is simpler (just the existence in the allowlist). If the list is enabled can be checked separately.

**Comment by webmaster128** [comment]
Thanks a lot. The work is integrated in #237. Happy if you look through that PR again.

**Pull request closed without merging**